xarray/core/dataarray.py

```diff
@@ -4414,13 +4420,13 @@ def to_iris(self) -> iris_Cube:
         return to_iris(self)
 
     @classmethod
-    def from_iris(cls, cube: iris_Cube) -> DataArray:
+    def from_iris(cls, cube: iris_Cube) -> Self:
         """Convert a iris.cube.Cube into an xarray.DataArray"""
         from xarray.convert import from_iris
 
         return from_iris(cube)
 
-    def _all_compat(self: T_DataArray, other: T_DataArray, compat_str: str) -> bool:
+    def _all_compat(self, other: T_DataArray, compat_str: str) -> bool:
```

Comment:
Here I'm not sure what the correct type is. A TypeVar in a method definition without using this TypeVar in the return type is useless.

The question is if other should be DataArray or Self?

Same with the next few changes below

Reply:
> A TypeVar in a method definition without using this TypeVar in the return type is useless.

Isn't it at least restricting it to `DataArray` and its subtypes? Or you mean useless relative to that?

Reply:
You can always assign a subclass to an argument.

Reply:
But `T_DataArray` is useless relative to having nothing there? Or it's useless relative to just having `DataArray`?

Reply:
> But `T_DataArray` is useless relative to having nothing there? Or it's useless relative to just having `DataArray`?

Ah ok, it is useless relative to just `DataArray`